hoodie-client/src/main/java/com/uber/hoodie/table/HoodieCopyOnWriteTable.java

```diff
@@ -174,40 +184,70 @@ public boolean isWorkloadProfileNeeded() {
   }
 
   protected Iterator<List<WriteStatus>> handleUpdateInternal(HoodieMergeHandle upsertHandle,
-      String commitTime, String fileLoc) throws IOException {
+      String commitTime, String fileLoc)
+      throws IOException {
     if (upsertHandle.getOldFilePath() == null) {
-      throw new HoodieUpsertException(
-          "Error in finding the old file path at commit " + commitTime + " at fileLoc: " + fileLoc);
+      throw new HoodieUpsertException("Error in finding the old file path at commit "
+          + commitTime + " at fileLoc: " + fileLoc);
     } else {
       AvroReadSupport.setAvroReadSchema(getHadoopConf(), upsertHandle.getSchema());
-      ParquetReader<IndexedRecord> reader = AvroParquetReader.builder(upsertHandle.getOldFilePath())
-          .withConf(getHadoopConf()).build();
+      ParquetReader<IndexedRecord> reader =
+          AvroParquetReader.builder(upsertHandle.getOldFilePath()).withConf(getHadoopConf())
+              .build();
+      final TaskContext sparkThreadTaskContext = TaskContext.get();
+      // Executor service used for launching writer thread.
+      final ExecutorService writerService = Executors.newFixedThreadPool(1);
       try {
-        IndexedRecord record;
-        while ((record = reader.read()) != null) {
-          // Two types of writes here (new record, and old record).
-          // We have already catch the exception during writing new records.
-          // But for old records, we should fail if any exception happens.
-          upsertHandle.write((GenericRecord) record);
-        }
-      } catch (IOException e) {
-        throw new HoodieUpsertException(
-            "Failed to read record from " + upsertHandle.getOldFilePath() + " with new Schema "
-                + upsertHandle.getSchema(), e);
+        // Used for buffering records which is controlled by HoodieWriteConfig#WRITE_BUFFER_LIMIT_BYTES.
+        final BufferedIterator<GenericRecord, GenericRecord> bufferedIterator =
+            new BufferedIterator<>(new ParquetReaderWrappedIterator(reader),
+                config.getWriteBufferLimitBytes(),
+                new GenericRecordBufferedIteratorPayloadFunction());
+        Future writerResult =
+            writerService.submit(
+                () -> {
+                  logger.info("starting hoodie writer thread");
```

Comment:
this code seems repeated, is there a way to modularize and share code? Also revisit logger statements to make then more contextualized?

Reply:
addressed.